### PR TITLE
[RLlib] Preparatory PR: Make EnvRunners use (enhanced) Connector API (#01: mostly cleanups and small fixes)

### PR DESCRIPTION
**Pull request by sven1977**
Preparatory PR: Make EnvRunners use (enhanced) Connector API (#01: mostly cleanups and small fixes)

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/algorithm.py [hunk 1]
```diff
@@ -1932,7 +1932,10 @@ def compute_actions(
         filtered_obs, filtered_state = [], []
         for agent_id, ob in observations.items():
             worker = self.workers.local_worker()
-            preprocessed = worker.preprocessors[policy_id].transform(ob)
+            if worker.preprocessors.get(policy_id) is not None:
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:1935]: This is a bug fix.

File: rllib/algorithms/algorithm_config.py [hunk 2]
```diff
@@ -319,26 +319,37 @@ def __init__(self, algo_class=None):
         # If not specified, we will try to auto-detect this.
         self._is_atari = None
 
+        # TODO (sven): Rename this method into `AlgorithmConfig.sampling()`
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:322]: Now that we are aiming for a the EnvRunner API as the default, we should rename/clarify some of these config settings and methods.

File: rllib/core/models/torch/encoder.py [hunk 3]
```diff
@@ -285,30 +285,31 @@ def __init__(self, config: RecurrentEncoderConfig) -> None:
             bias=config.use_bias,
         )
 
+        self.state_in_out_spec = {
```
Comment (sven1977) [inline rllib/core/models/torch/encoder.py:288]: Simplified (repetitive) code.

File: rllib/env/multi_agent_episode.py [hunk 4]
```diff
@@ -212,75 +212,75 @@ def get_observations(
 
         return self._getattr_by_index("observations", indices, global_ts)
 
-    def get_actions(
+    def get_infos(
```
Comment (sven1977) [inline rllib/env/multi_agent_episode.py:215]: Reordered:
1) obs, infos (<- env.reset data)
2) action, reward, terminated/truncated (<- other env.step results)
3) extra model outs

File: rllib/env/single_agent_env_runner.py [hunk 5]
```diff
@@ -41,56 +41,48 @@ def __init__(self, config: "AlgorithmConfig", **kwargs):
...
         self.worker_index: int = kwargs.get("worker_index")
 
+        # Create the vectorized gymnasium env.
+
         # Register env for the local context.
         # Note, `gym.register` has to be called on each worker.
-        gym.register(
-            "custom-env-v0",
-            partial(
+        if (
```
Comment (sven1977) [inline rllib/env/single_agent_env_runner.py:48]: Bug fix.

File: rllib/evaluation/worker_set.py [hunk 6]
```diff
@@ -690,6 +690,9 @@ def foreach_worker(
         if local_worker and self.local_worker() is not None:
             local_result = [func(self.local_worker())]
 
+        if not self.__worker_manager.actor_ids():
```
Comment (sven1977) [inline rllib/evaluation/worker_set.py:693]: Shortcut for local-worker only case.

File: rllib/tuned_examples/appo/multi-agent-cartpole-crashing-restart-env-appo.yaml [hunk 7]
```diff
@@ -30,7 +30,7 @@ multi-agent-cartpole-crashing-appo:
         # Switch on resiliency for failed sub environments (within a vectorized stack).
         restart_failed_sub_environments: true
 
-        # Switch on evaluation workers being managed by AsyncRequestsManager object.
+        # Switch on asynchronous handling of evaluation workers.
```
Comment (sven1977) [inline rllib/tuned_examples/appo/multi-agent-cartpole-crashing-restart-env-appo.yaml:33]: `AsyncRequestsManager` doesn't exist anymore.

File: rllib/utils/spaces/space_utils.py [hunk 8]
```diff
@@ -205,6 +205,56 @@ def flatten_to_single_ndarray(input_):
     return input_
 
 
+@DeveloperAPI
```
Comment (sven1977) [inline rllib/utils/spaces/space_utils.py:208]: Very useful new utility. Inverse of already existing `unbatch` utility.

File: rllib/utils/spaces/space_utils.py [hunk 9]
```diff
@@ -205,6 +205,56 @@ def flatten_to_single_ndarray(input_):
...
+def batch(list_of_structs, individual_items_already_have_batch_1: bool = False):
```
Comment (kouroshHakha) [inline rllib/utils/spaces/space_utils.py:209]: data types please (for input and output)

Comment (kouroshHakha) [reply rllib/utils/spaces/space_utils.py:209]: can we have unittest of this ?

File: rllib/utils/spaces/space_utils.py [hunk 10]
```diff
@@ -205,6 +205,56 @@ def flatten_to_single_ndarray(input_):
...
+
+    for item in list_of_structs:
+        flattened_item = tree.flatten(item)
+        # Create the main list, in which each slot represents one leaf in the (nested)
+        # struct. Each slot holds a list of batch values.
+        if flat is None:
+            flat = [[] for _ in range(len(flattened_item))]
+        for i, value in enumerate(flattened_item):
+            flat[i].append(value)
+
```
Comment (kouroshHakha) [inline rllib/utils/spaces/space_utils.py:250]: add: 
```
    if item is None:
        raise ValueError("Input list_of_structs does not contain valid structs.")

```

File: rllib/utils/spaces/space_utils.py [hunk 11]
```diff
@@ -205,6 +205,56 @@ def flatten_to_single_ndarray(input_):
...
+        individual_items_already_have_batch_1: True, if the individual items in
+            `list_of_structs` already have a batch dim (of 1). In this case, we will
+            concatenate (instead of stack) at the end.
+
+    Returns:
+        The struct of component batches. Each leaf item
+        in this struct represents the batch for a single component
+        (in case struct is tuple/dict). Alternatively, a simple batch of
+        primitives (non tuple/dict) might be returned.
+    """
```
Comment (kouroshHakha) [inline rllib/utils/spaces/space_utils.py:239]: add
```
    if not list_of_structs:
        raise ValueError("Input list_of_structs is empty.")

```

**Review by kouroshHakha — APPROVED** [review]

File: rllib/core/models/torch/encoder.py (hunk 3 above)
Comment (sven1977) [reply rllib/core/models/torch/encoder.py:288]: done

File: rllib/utils/spaces/space_utils.py (hunk 10 above)
Comment (sven1977) [reply rllib/utils/spaces/space_utils.py:250]: done

File: rllib/utils/spaces/space_utils.py (hunk 11 above)
Comment (sven1977) [reply rllib/utils/spaces/space_utils.py:239]: done

File: rllib/utils/spaces/space_utils.py (hunk 9 above)
Comment (sven1977) [reply rllib/utils/spaces/space_utils.py:209]: done and done

Comment (sven1977) [reply rllib/utils/spaces/space_utils.py:209]: also enhanced the docstring to make the example and explanations more clear.

**Comment by sven1977** [comment]
Thanks for the review @kouroshHakha ! Waiting for tests to pass ...

**Pull request merged**